### ISSUE TITLE
Add undefined to tuple index signature, or reject keys that are not "keyof tuple"

### ISSUE THREAD

**Issue by lonewarrior556** [microsoft/TypeScript#38779]
## Search Terms

Tuple index signature

## Suggestion
I am opening up this issue as Seperate from https://github.com/microsoft/TypeScript/issues/13778 since I agree with the current interpretation of the array 
```ts
declare var strArray: string[] 
var a : strArray[1000 as number]  // ok; a is string
```

but this fails in conjunction with tuples
```ts
declare var strTuple: [number, number] 
var a : strTuple[1000 as number]  // absolutely not ok; a is string; 
```

```
var a : strTuple[10 as number] 
``` 
should either be
- `string | undefined`
or
- should give a type error similar to: 
```
var a = { 0:  10, 1: 20} as const;
a[2 as number]; // type 'number' can't be used to index type
```

## Use Cases

// Easy fix for for loops
```
for (let i = 0 as 0 | 1; i < 2; i++) { 
  var defString = strTuple[i] // type string
  var j = i + someNumber;
  var not_defstring = strTuple[j] //lets you know you need to be careful
}
```
You could use `keyof typeof strTuple` instead of `0 |` 1 when you fix this https://github.com/microsoft/TypeScript/issues/27995

Given that strictNull checks still allow typed arrays to do this:
```
var notString1 = aa[Infinity]; // no error, not undefined
var notString2 = aa[NaN]; // no error, not undefined
```
It would be nice to have a mechanism that could allow you to declare arrays with better type safety. 

## Checklist

My suggestion meets these guidelines:

* [x] This wouldn't be a breaking change in existing TypeScript/JavaScript code
* [x] This wouldn't change the runtime behavior of existing JavaScript code
* [x] This could be implemented without emitting different JS based on the types of the expressions
* [x] This isn't a runtime feature (e.g. library functionality, non-ECMAScript syntax with JavaScript output, etc.)
* [x] This feature would agree with the rest of [TypeScript's Design Goals](https://github.com/Microsoft/TypeScript/wiki/TypeScript-Design-Goals).

**Comment by thw0rted** [comment]
If I understand your issue correctly, you want `[T,U,V][number]` to type as `T | U | V | undefined` instead of just `T | U | V`.  Is that right?  I don't see how that's any different from / better than #13778.

>  This wouldn't be a breaking change in existing TypeScript/JavaScript code

This is not true.

````ts
declare const two: [string, string];
const oneOfthem = two[Math.floor(Math.random()*2)];
oneOfThem.charAt(0);
````

Before your proposed change, `oneOfThem` types as `string`, but after it would type as `string | undefined`.  All the arguments that have been made against #13778 would then apply -- "but I checked that the index stays in-bounds!", etc.

**Comment by lonewarrior556** [comment]
> If I understand your issue correctly, you want `[T,U,V][number]` to type as `T | U | V | undefined` instead of just `T | U | V`. Is that right?
( yes, either that or throw // type 'number' can't be used to index type)

>  I don't see how that's any different from / better than #13778.
tuples were release much later and much more limited in use.. so unlike #13778 this would not ` be the single most breaking change in any project I've ever followed.`


> 
> > This wouldn't be a breaking change in existing TypeScript/JavaScript code
> 
> This is not true.

( you are right,, but it would be a minor breaking change,
 And honestly because the tuple `[string]` can accept `[0 as const]` but not `[1 as const]` mean it cannot accept `number` 
so this is closer to a bug fix )

> ```ts
> declare const two: [string, string];
> const oneOfthem = two[Math.floor(Math.random()*2)];
> oneOfThem.charAt(0);
> ```

Unfortuanly TS does not know `Math.floor(1 as const)` is `1` you expect it to?


> Before your proposed change, `oneOfThem` types as `string`, but after it would type as `string | undefined`. All the arguments that have been made against #13778 would then apply -- "but I checked that the index stays in-bounds!", etc.

**Comment by thw0rted** [comment]
I still don't see this as a "minor" breaking change, but I'm not sure it's a big enough deal to merit a lot of debate.  I make a point of not explicitly indexing arrays unless I absolutely have to -- I use for/of or forEach whenever possible -- so really, tuples should be no different.  The large majority of existing tuple usage should be indexed with a literal number, right?  But those cases already work, because the system is smart enough to treat `tuple[1]` as `tuple[1 as const]` rather than `tuple[1 as number]`.  So already, this approach is trying to solve a problem that shouldn't happen often.  Right?

**Comment by lonewarrior556** [comment]
All I am saying is this

```
const t0: [string][0] = 'hello' // ok
const t1: [string][1] = 'hello' // Type '"hello"' is not assignable to type 'undefined'.

// therefore this is already true
const t2: [string][0|1] // string | undefined 

// How can you justify this behavior?
const t3: [string][number] // string 
```

Yes, `for of`/ `forEach` loops are better (although the still don't protect you from spares arrays)
But there are uses for using array indices

[this](https://github.com/microsoft/TypeScript/issues/13778#issuecomment-626850178) was what brought me to the other thread:

> ```ts
> var notString1 = aa[Infinity]; // no error, not undefined
> var notString2 = aa[NaN]; // no error, not undefined
> ```
> 
> Which originally caused my runtime error as a `number` became a `NaN`, then returned an `undefined` from a tuple..
> All of which was type safe

So objectively this is an issue that does happen.

**Comment by thw0rted** [comment]
I wonder if maybe what you need for your specific issue is actually https://github.com/microsoft/TypeScript/pull/29317 ?  If negation existed, you could declare your index as `type FineiteNumber = number & (not Infinity) & (not -Infinity) & (not NaN)`, then e.g. `parseInt` would return `number` and you'd be reminded to type-guard down to `FiniteNumber` by checking for those edge cases before using it as a tuple index.

Obviously this does nothing for `[string, string][5]` but that's a different class of error and frankly you're a lot less likely to be in a position where the type checker could notice that it's a 5.
